### PR TITLE
storage: Some typing fun

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

**Comment by mvollmer** [comment]
Somehow TS doesn't allow this:

```
const items = [];
items.push("123");
```

TS infers "never[]" as the type for "items" and doesn't allow pushing strings onto it.

But @allisonkarlitskaya says this should work and TS should infer "string[]" as the type, and it does in her experiments.

File: tsconfig.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
         "moduleResolution": "node",
         "noEmit": true,  // we only use `tsc` for type checking
         "strict": true,
+        "noImplicitAny": false,
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply tsconfig.json:26]: > Can you please split this bit out as a separate commit with some verbiage about the impact and trade-offs?

I hope we don't need to do this globally.

> Does this apply to *.ts as well? (which we should keep strict), or just to *.js (then that'd be nice)

It applies to all files, unfortunately. I'll see how to switch this off for individual files.

File: pkg/lib/cockpit.d.ts [hunk 2]
```diff
@@ -188,6 +188,10 @@ declare module 'cockpit' {
         href: string;
         go(path: Location | string, options?: { [key: string]: string }): void;
         replace(path: Location | string, options?: { [key: string]: string }): void;
+
+        encode(path: string | Array<string>,
+               options? : { [name: string]: string | Array<string> },
```
Comment (allisonkarlitskaya) [inline pkg/lib/cockpit.d.ts:193]: You can also write `Record<>` for that, I think?

File: tsconfig.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply tsconfig.json:26]: A setting for individual files would be OK.  Some kind of comment at the top or so?

File: pkg/lib/cockpit.d.ts (hunk 2 above)
Comment (mvollmer) [reply pkg/lib/cockpit.d.ts:193]: I was just copy/pasting... :-)

File: pkg/storaged/storage-controls.tsx [hunk 3]
```diff
@@ -47,28 +47,26 @@ const _ = cockpit.gettext;
...
 
-class StorageControl extends React.Component {
-    render() {
-        const excuse = this.props.excuse;
-        if (!client.superuser.allowed)
-            return <div />;
+const StorageControl = ({ excuse, excuse_placement, content } :
+                        { excuse, excuse_placement?, content }) => {
+    if (!client.superuser.allowed)
+        return <div />;
```
Comment (cockpituous) [inline pkg/storaged/storage-controls.tsx:53]: This added line is not executed by any test.

File: pkg/storaged/storage-controls.tsx [hunk 4]
```diff
@@ -120,13 +119,13 @@
                                        onClick={checked(onClick, setSpinning)}
                                        variant={kind || "secondary"}
                                        isDisabled={!!excuse || (spinner && spinning)}
-                                       isLoading={spinner ? spinning : undefined}>
+                                       isLoading={spinner ? spinning : false}>
```
Comment (cockpituous) [inline pkg/storaged/storage-controls.tsx:122]: This added line is not executed by any test.

File: pkg/storaged/storage-controls.tsx [hunk 5]
```diff
@@ -141,43 +140,36 @@
...
-            }
-
-            self.setState({ promise, promise_goal_state: val });
+    function _onChange(_event, val) {
+        const promise = onChange(val);
+        if (promise) {
+            promise.catch(error => {
+                dialog_open({
+                    Title: _("Error"),
+                    Body: error.toString()
```
Comment (cockpituous) [inline pkg/storaged/storage-controls.tsx:153]: These 4 added lines are not executed by any test.

**Comment by mvollmer** [comment]
> Somehow TS doesn't allow this:
> 
> ```
> const items = [];
> items.push("123");
> ```
> 
> TS infers "never[]" as the type for "items" and doesn't allow pushing strings onto it.
> 
> But @allisonkarlitskaya says this should work and TS should infer "string[]" as the type, and it does in her experiments.

This is caused by the version of "tsc". I had 5.1.3  installed from Fedora repos, which shows this behavior. I also have node_modules/.bin/tsc which is 5,6,3, and infers "any[]"...

**Comment by mvollmer** [comment]
I'll try this again.

**Comment by mvollmer** [comment]
> > Somehow TS doesn't allow this:
> > ```
> > const items = [];
> > items.push("123");
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > TS infers "never[]" as the type for "items" and doesn't allow pushing strings onto it.
> > But @allisonkarlitskaya says this should work and TS should infer "string[]" as the type, and it does in her experiments.
> 
> This is caused by the version of "tsc". I had 5.1.3 installed from Fedora repos, which shows this behavior. I also have node_modules/.bin/tsc which is 5,6,3, and infers "any[]"...

But ALSO, noImplicitAny: false seems to change tsc 5.6.3 into inferring "never[]"... Or I am losing my mind.

**Pull request closed without merging**